core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -32,33 +79,52 @@
 @Experimental
 public interface NonBlockingStore<K, V> {
 
+   /**
+    * Enumeration defining the various characteristics of the underlying store to communicate what features it may
+    * or may not support.
+    */
    enum Characteristic {
       /**
        * Whether this cache can be shared between multiple nodes. An example would be an external system, such as
-       * a database.
+       * a database. This characteristic is here solely for validation of the store configuration.
        */
       SHAREABLE,
       /**
        * If this store only supports being read from.  Any write based operations will never be invoked on this store.
+       * No optional methods map to this characteristic. The {@link #write(int, MarshallableEntry)},
+       * {@link #delete(int, Object)}, {@link #bulkWrite(int, Publisher)}, and {@link #bulkDelete(int, Publisher)} methods
+       * will not be invoked on a store that has this characteristic.
```

Comment:
Hmm this is making me thinking that even write() etc should have no-op default methods that throw a `UnsupportedOperationException`. If I'm creating a READ_ONLY store why should I have to implement the method?

Reply:
I had debated about that, however I felt that this case was rare enough that it is simpler to ensure they implement it. Because you can say the same thing with load in which case we wouldn't have many method to implement.

The other thing is we can always add a default later, however we can't remove default once it is there.

Reply:
Yeah that's a good point. We can always address this if required later on.